src/bin/lychee/stats.rs

```diff
@@ -1,65 +1,137 @@
+use pad::{Alignment, PadStr};
 use serde::{Deserialize, Serialize};
 
 use std::{
-    collections::HashSet,
+    collections::HashMap,
     fmt::{self, Display},
 };
 
-use lychee::{Response, Status::*, Uri};
+use lychee::{collector::Input, Response, Status::*, Uri};
+
+// Maximum padding for each entry in the final statistics output
+const MAX_PADDING: usize = 20;
 
 #[derive(Serialize, Deserialize)]
 pub struct ResponseStats {
     total: usize,
     successful: usize,
-    failures: HashSet<Uri>,
-    timeouts: HashSet<Uri>,
-    redirects: HashSet<Uri>,
-    excludes: HashSet<Uri>,
-    errors: HashSet<Uri>,
+    failures: usize,
+    timeouts: usize,
+    redirects: usize,
+    excludes: usize,
+    errors: usize,
+    fail_map: HashMap<Input, Vec<Uri>>,
 }
 
 impl ResponseStats {
     pub fn new() -> Self {
+        let fail_map = HashMap::new();
         ResponseStats {
             total: 0,
             successful: 0,
-            failures: HashSet::new(),
-            timeouts: HashSet::new(),
-            redirects: HashSet::new(),
-            excludes: HashSet::new(),
-            errors: HashSet::new(),
+            failures: 0,
+            timeouts: 0,
+            redirects: 0,
+            excludes: 0,
+            errors: 0,
+            fail_map,
         }
     }
 
     pub fn add(&mut self, response: Response) {
         self.total += 1;
-        let uri = response.uri;
-        if !match response.status {
-            Failed(_) => self.failures.insert(uri),
-            Timeout => self.timeouts.insert(uri),
-            Redirected => self.redirects.insert(uri),
-            Excluded => self.excludes.insert(uri),
-            Error(_) => self.errors.insert(uri),
-            _ => false,
-        } {
-            self.successful += 1;
+        match response.status {
+            Failed(_) => self.failures += 1,
+            Timeout(_) => self.timeouts += 1,
+            Redirected(_) => self.redirects += 1,
+            Excluded => self.excludes += 1,
+            Error(_) => self.errors += 1,
+            _ => self.successful += 1,
         }
+
+        if matches!(
+            response.status,
+            Failed(_) | Timeout(_) | Redirected(_) | Error(_)
+        ) {
+            let fail = self.fail_map.entry(response.source).or_default();
+            fail.push(response.uri);
```

Comment:
This is a bit sad. It means that we don't show the response status in the output yet. That's because `Status` is an enum which wraps around `http::StatusCode`, which is not `Serialize`. Maybe we find a way to do that in the future. In the worst case we can push a String with the pre-formatted status code + the uri. I'd like to avoid that, though to keep the machine-readable output clean and parseable without grep/cut/...

Reply:
Fixed. We now use the `Response` type with full status/error output.